### ISSUE TITLE
[Feature Request] PEP 604 Support

### ISSUE THREAD

**Issue by dycw** [beartype/beartype#71]
Hi @leycec , thanks for a great library!

Upon discovered of [this](https://github.com/beartype/beartype/issues/56#issuecomment-938713933) decorator, I thought I'd give `beartype` a run in my codebase -- I am a heavy user of `dataclass`es. And, of course, a heavy user of runtime-checks!

After the proof-of-concept work, everything still works. But I did notice:

- the `from future import __annotations__` had to be removed (I'm a 3.9 user). Is this required by your users?
- the *initial* performance was poorer -- I'd continue to move the handcrafted runtime checks into `Is[]`. I'd report back later on this.

**Comment by leycec** [comment]
> ...thanks for a great library!

Oh, you're *most* welcome. Let's make it even greater, everyone! :hugs: 

> everything still works

This calls for celebratory gunpowder-filled explosives. :fireworks: :boom: :fireworks: 

> the `from future import __annotations__` had to be removed (I'm a 3.9 user).

**Yikes.** That's unexpected. ``beartype`` explicitly supports [PEP 563](https://www.python.org/dev/peps/pep-0563) in all its hideous glory and shame. In fact, we pride ourselves on fielding the best, fastest, and most robust PEP 563 resolver – unlike our more popular forebears who shall remain nameless because they're *really* super nice. <sup>*cough* `pydantic` *cough*</sup>

Did something specifically blow up on you? If so, we'll fix that post-haste. Hit me repeatedly with your best traceback and I'll push out a new stable release resolving that for everyone within four picoyears ± twelve nanodays.

> the initial performance was poorer

**Yup.** That's expected, sadly. I know, right? [PEP 563](https://www.python.org/dev/peps/pep-0563): the busted standard that was so busted it had to be forcefully removed from Python 3.10 by jackhammers at dawn before anyone noticed.

Both PEP 563 and [PEP 569, PEP 563's possible replacement](https://www.python.org/dev/peps/pep-0649) substantially reduce performance. If Python's Steering Committee ever forces a future version of Python to unconditionally enable either of these PEPs, both ``beartype`` and all other runtime type checkers **will become much, much slower** – possibly prohibitively slower.

If this concerns you, please consider posting to the [official `python-dev` mailing list](https://mail.python.org/archives/list/python-dev@python.org/). :disappointed:

**Comment by TeamSpen210** [comment]
Well, latest news is that the Steering Council has [deferred accepting](https://mail.python.org/archives/list/python-dev@python.org/thread/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/) either of those PEPs, and are requesting more input from all stakeholders.

**Comment by leycec** [comment]
`</phew>`

Feels like this is PEP Bullet Hell and [we're still at the first Ikaruga boss](https://www.youtube.com/watch?v=eelfcFaKEZ0). These PEPs will rise from the grave like shambolic standard zombies. When they do, @beartype will be there with a fistful of timings and an unruly mob waving incoherent placards.

If only Cisco Systems would [publicly release their runtime type-checking profiling suite](https://github.com/beartype/beartype/issues/58#issuecomment-940100279). Apparently, @beartype is faster than Cythonized ``pydantic`` by two orders of magnitude. How is this possible!?!?

**Comment by dycw** [comment]
Hi @leycec , let me try work out a minimal example.

**Comment by dycw** [comment]
Hi @leycec , I remember now; please see [this repo](https://github.com/dycw/beartype-annotations).

**Comment by leycec** [comment]
**Wow.** Thanks a Canadian-standard metric ton for [`beartype-annotations`](https://github.com/dycw/beartype-annotations), Derek. Yet again, you're *absolutely* right; that's shameful badness on `beartype`'s part and make no mistake. :sweat: 


## Well, This Is Awkward

I'm actually kinda shocked that `int | None` is valid Python. I mean, `int | type(None)`, *sure* – because `type(None)` is a type. `None` isn't. That would be like allowing `int | 0x42`, you know? But Python clearly wants `None` to be transparently usable as a type in type unions, even if that doesn't really make semantic or syntactic sense and violates everything good and sane left in the multiverse.

At least Python internally reduces `int | None` to `int | type(None)` at runtime:

``` python
>>> (int | type(None)).__args__
(int, NoneType)
```

Very well, Guido: *you win this round.*

That must be yet another inscrutable small-font footnote embedded at the end of [PEP 604](https://www.python.org/dev/peps/pep-0604) where they knew no one would look. When you're speed-reading these things with one eye while grinding demon fusions in *Persona 5 Royal* with the other, you're bound to royally screw up both. So much "Urgh!"

## We May not Be Able to Fix This

*uhoh.* You may have discovered a critical bug in Python itself. I can trivially confirm that `eval('int | None')` behaves as expected. This can only mean that `from __future__ import annotations` (i.e., PEP 563) is erroneously stringifying `int | None` into `int | NoneType`. Since `NoneType` doesn't actually exist anywhere, `eval('int | NoneType')` raises the exception you're seeing.

In other words, we may need to escalate this to the Python bug tracker. `</gulp>`

**Comment by leycec** [comment]
**Oh, oh, oh.** I finally exhausted five spare brain cells (which now seem to be comatose) on this. Phew! Hope I didn't need those cells. On the bright side, Python ain't broke. On the dark side, both of us are:

* `foo | bar` only works under **Python ≥ 3.10**. Python 3.9 is right out. That explains why this failed hard for you: "I'm a 3.9 user." ←<sub>*problem statement*</sub>
* `foo | bar` still doesn't work under Python ≥ 3.10 with `beartype 0.9.1`. ←<sub>*problem statement*</sub>

We foolishly speedread ["PEP 604 -- Allow writing union types as X | Y"](https://www.python.org/dev/peps/pep-0604) to mean that:

```python
foo | bar is typing.Union[foo, bar]
```

In fact, PEP 484 unions (e.g., `typing.Union[...]`) and PEP 604 unions (e.g., `foo | bar`) hate each other's standardized guts and are pretending the other guy no longer exists. Yes, PEP 484 and 604 unions are ghosting each another. What is this – Tinder all over again!?!?

Of course, this is insane. In a sane world, PEP 484 and 604 unions would have been implemented in terms of the same underlying primitives. This is not that world. We're on the bad timeline, guys. We now need to explicitly support both PEP 484 and 604 unions as totally distinct things that absolutely have no semantic or syntactic relation to one another. `</facepalm>`

Our upcoming `beartype 0.10.0` release will explicitly support New Wave PEP 604 unions. Until then, please continue using [Ye Ol' PEP 484 unions](https://www.python.org/dev/peps/pep-0484/#union-types). The collective sadness of humanity grows by the day.

**Comment by dycw** [comment]
This is a sad world indeed. But somewhere someone told me we're working in the

> lawless world of dynamically-typed Python

and it wouldn't be lawless without some lawlessness, right?

Damn, these unions are crazy. Can't wait for the new release - thanks @leycec !

**Comment by leycec** [comment]
**Yo, homie!** Crazy union supported just dropped. Now you too can leverage the dark power of the `|` operator to satisfy all your succinct typing needs. Wow coworkers! Wow the wifey! Wow even that twenty-year stray cat with mange you took in against your University roommate's wishes. You showed him, didn't you? Now, show everyone else.

All this and more is coming up in... `beartype 0.10.0`. Activate the thunderbolts and lightning, Zeus. :cloud_with_lightning:

**Comment by dycw** [comment]
@leycec すごい! 

The codebase is going to be so happy for this. Really, amazing stuff!

**Comment by leycec** [comment]
でしょうね！

*...heh.* I just love sounding like a polite Japanese housewife. Gotta rock that 東京方言 (Tokyo-ben; Tokyo dialect) for those nostalgic feels. Oh, boy! All this hiragana got me fired up to repeatedly beat up uncouth Japanese gangsters as Kiryu-chan in 龍が如く極 (Yakuza Kiwami) this weekend. **Dis gonna be fun.**

![kiryu-chan](https://steamuserimages-a.akamaihd.net/ugc/973245563739983916/AD9D9E840C4B09D0BE29ADA77E7C031643764F13/?imw=1024&imh=576&ima=fit&impolicy=Letterbox&imcolor=%23000000&letterbox=true)
<sup>I feel this way about PEP 640.</sup>
